src/org/traccar/reports/ReportUtils.java

```diff
@@ -250,100 +255,79 @@ private static boolean isMoving(ArrayList<Position> positions, int index,
+                    startTripIndex = -1;
+                }
+            }
+            if (startTripIndex != -1 && startParkingIndex != -1) {
+                result.add(calculateTrip(positions, startTripIndex, startParkingIndex, ignoreOdometer));
+            }
+        }
+        return result;
+    }
+
+    public static Collection<StopReport> detectStops(Collection<Position> positionCollection, TripsConfig tripsConfig,
+            boolean ignoreOdometer, double speedThreshold) {
+        Collection<StopReport> result = new ArrayList<>();
+
+        ArrayList<Position> positions = new ArrayList<>(positionCollection);
+        if (positions != null && !positions.isEmpty()) {
+            MotionEventHandler  motionHandler = new MotionEventHandler(tripsConfig);
+            DeviceState deviceState = new DeviceState();
+            deviceState.setMotionState(isMoving(positions, 0, tripsConfig, speedThreshold));
+            int startTripIndex = -1;
+            int startParkingIndex = deviceState.getMotionState() ? -1 : 0;
+            for (int i = 0; i < positions.size(); i++) {
+                positions.get(i).set(Position.KEY_MOTION, isMoving(positions, i, tripsConfig, speedThreshold));
+                Event event = motionHandler.updateMotionState(deviceState, positions.get(i));
+                if (startParkingIndex == -1 && deviceState.getMotionState()
+                        && deviceState.getMotionPosition() != null) {
+                    startParkingIndex = i;
+                    startTripIndex = -1;
+                }
+                if (!deviceState.getMotionState()) {
+                    if (startTripIndex == -1) {
+                        startTripIndex = i;
+                    } else if (deviceState.getMotionPosition() == null) {
+                        startTripIndex = -1;
                     }
+                }
+                if (startParkingIndex != -1 && startTripIndex != -1 && event != null && deviceState.getMotionState()) {
+                    result.add(calculateStop(positions, startParkingIndex, startTripIndex));
                     startParkingIndex = -1;
                 }
             }
-            if (result.isEmpty() && !trips) {
-                int end = isMoving && !tripsConfig.getGreedyParking()
-                        ? Math.max(endParkingIndex, previousEndParkingIndex) : positions.size() - 1;
-                long parkingDuration = positions.get(end).getFixTime().getTime()
-                        - positions.get(previousStartParkingIndex).getFixTime().getTime();
-                if (parkingDuration >= tripsConfig.getMinimalParkingDuration()) {
-                    result.add(calculateStop(positions, previousStartParkingIndex, end));
-                }
+            if (startParkingIndex != -1) {
+                result.add(calculateStop(positions, startParkingIndex,
+                        startTripIndex != -1 ? startTripIndex : positions.size() - 1));
```

Comment:
If we detected too short trip add parking, or assume parking till the end of period.